src/Maestro/Maestro.DataProviders/DarcRemoteFactory.cs

```diff
@@ -2,37 +2,169 @@
+            using (logger.BeginScope($"Getting remote for repo {repoUrl}."))
+            {
+                // Normalize the url with the AzDO client prior to attempting to
+                // get a token. When we do coherency updates we build a repo graph and
+                // may end up traversing links to classic azdo uris.
+                string normalizedUrl = AzureDevOpsClient.NormalizeUrl(repoUrl);
+                Uri normalizedRepoUri = new Uri(normalizedUrl);
+                // Look up the setting for where the repo root should be held.  Default to empty,
+                // which will use the temp directory.
+                string temporaryRepositoryRoot = Configuration.GetValue<string>("DarcTemporaryRepoRoot", null);
+                if (string.IsNullOrEmpty(temporaryRepositoryRoot))
+                {
+                    temporaryRepositoryRoot = TempFiles.GetFilePath("repos");
+                }
+                IGitRepo gitClient;
+
+                long installationId = await Context.GetInstallationId(normalizedUrl);
+
+                await ExponentialRetry.RetryAsync(
+                    async () => await EnsureLocalGit(logger),
+                    ex => logger.LogError(ex, $"Failed to install git to local temporary directory."),
+                    ex => true);
+
+                switch (normalizedRepoUri.Host)
+                {
+                    case "github.com":
+                        if (installationId == default)
+                        {
+                            throw new GithubApplicationInstallationException($"No installation is avaliable for repository '{normalizedUrl}'");
+                        }
+                        gitClient = new GitHubClient(GitExecutable, await GitHubTokenProvider.GetTokenForInstallationAsync(installationId),
+                            logger, temporaryRepositoryRoot, Cache.Cache);
+                        break;
+                    case "dev.azure.com":
+                        gitClient = new AzureDevOpsClient(GitExecutable, await AzureDevOpsTokenProvider.GetTokenForRepository(normalizedUrl),
+                            logger, temporaryRepositoryRoot);
+                        break;
+                    default:
+                        throw new NotImplementedException($"Unknown repo url type {normalizedUrl}");
+                };
+
+                return new Remote(gitClient, new MaestroBarClient(Context, KustoClientProvider), logger);
+            }
+        }
+
+        /// <summary>
+        ///     Download and install git to the a temporary location.
+        ///     Git is used by DarcLib, and the Service Fabric nodes do not have it installed natively.
+        ///     
+        ///     The file is assumed to be on a public endpoint.
+        ///     We return the git client executable so that this call may be easily wrapped in RetryAsync
+        /// </summary>
+        public async Task<string> EnsureLocalGit(ILogger logger)
```

Comment:
What is the git executable used for here? I don't think the dependency graph paths utilize it at all.

Reply:
Looks like it does utilize the git executable:
- https://github.com/dotnet/arcade-services/blob/master/src/Microsoft.DotNet.Darc/src/DarcLib/Models/Darc/DependencyGraph.cs#L514
- https://github.com/dotnet/arcade-services/blob/master/src/Microsoft.DotNet.Darc/src/DarcLib/Models/Darc/DependencyGraph.cs#L846
- https://github.com/dotnet/arcade-services/blob/master/src/Microsoft.DotNet.Darc/src/DarcLib/Actions/Remote.cs#L1039
- https://github.com/dotnet/arcade-services/blob/master/src/Microsoft.DotNet.Darc/src/DarcLib/Actions/Remote.cs#L1081

Reply:
It does need a git client class, but I'm pretty sure it doesnt' need git.exe. I think the get dependencies call actually just makes a GitHub or AzDO GET call.

Basically, git.exe has bitten us in the past with large numbers ending up on the box, so the fewer third party downloads we do the better.

You can probably refactor this a bit so that the places that actually need git.exe have a separate check for the availability of git.exe.

Reply:
Got it. Let me look look into this.

Reply:
Yeah, it works!